File: AtividadesBFS/ex2.py
from collections import deque

# Grafo de rotas de ônibus de San Francisco (não-direcionado)
# Cada par de paradas compartilha uma aresta (uma linha de ônibus direta)
grafo = {
    "Twin Peaks":         ["Castro", "West Portal"],
    "Castro":             ["Twin Peaks", "Market & Church", "Noe Valley"],
    "West Portal":        ["Twin Peaks", "Forest Hill", "Sunset District"],
    "Market & Church":    ["Castro", "Civic Center", "Mission District"],
    "Noe Valley":         ["Castro", "Mission District"],
    "Forest Hill":        ["West Portal", "Civic Center"],
    "Sunset District":    ["West Portal", "Richmond District"],
    "Civic Center":       ["Market & Church", "Forest Hill", "Downtown"],
    "Mission District":   ["Market & Church", "Noe Valley", "Downtown"],
    "Richmond District":  ["Sunset District", "Golden Gate Park"],
    "Downtown":           ["Civic Center", "Mission District", "Fisherman's Wharf"],
    "Golden Gate Park":   ["Richmond District", "Presidio"],
    "Fisherman's Wharf":  ["Downtown", "Presidio"],
    "Presidio":           ["Golden Gate Park", "Fisherman's Wharf", "Golden Gate Bridge"],
    "Golden Gate Bridge":  ["Presidio"],
}

def bfs_menor_caminho(grafo, origem, destino):
    if origem == destino:
        return 0, [origem]

    fila = deque()
    fila.append((origem, [origem]))   # (parada_atual, caminho_percorrido)
    visitados = {origem}

    while fila:
        atual, caminho = fila.popleft()

        for vizinho in grafo.get(atual, []):
            if vizinho not in visitados:
                novo_caminho = caminho + [vizinho]

                if vizinho == destino:
                    etapas = len(novo_caminho) - 1
                    return etapas, novo_caminho

                visitados.add(vizinho)
                fila.append((vizinho, novo_caminho))

    return -1, []   # destino inalcançável

# ── Execução ──────────────────────────────────────────────────
origem  = "Twin Peaks"
destino = "Golden Gate Bridge"

etapas, caminho = bfs_menor_caminho(grafo, origem, destino)

print(f"De: {origem}")
print(f"Para: {destino}")
print(f"Etapas (arestas): {etapas}")
print(f"Caminho: {' → '.join(caminho)}")

File: AtividadesBFS/ex2.js
// AVISO: Não sei fazer isso em JavaScript.
// Peguei o mesmo código Python e traduzi para JS mantendo a mesma lógica.

const grafo = {
  "Twin Peaks":        ["Castro", "West Portal"],
  "Castro":            ["Twin Peaks", "Market & Church", "Noe Valley"],
  "West Portal":       ["Twin Peaks", "Forest Hill", "Sunset District"],
  "Market & Church":   ["Castro", "Civic Center", "Mission District"],
  "Noe Valley":        ["Castro", "Mission District"],
  "Forest Hill":       ["West Portal", "Civic Center"],
  "Sunset District":   ["West Portal", "Richmond District"],
  "Civic Center":      ["Market & Church", "Forest Hill", "Downtown"],
  "Mission District":  ["Market & Church", "Noe Valley", "Downtown"],
  "Richmond District": ["Sunset District", "Golden Gate Park"],
  "Downtown":          ["Civic Center", "Mission District", "Fisherman's Wharf"],
  "Golden Gate Park":  ["Richmond District", "Presidio"],
  "Fisherman's Wharf": ["Downtown", "Presidio"],
  "Presidio":          ["Golden Gate Park", "Fisherman's Wharf", "Golden Gate Bridge"],
  "Golden Gate Bridge": ["Presidio"],
};

function bfsMenorCaminho(grafo, origem, destino) {
  if (origem === destino) return { etapas: 0, caminho: [origem] };

  const fila = [[origem, [origem]]]; // equivale ao deque do Python
  const visitados = new Set([origem]);

  while (fila.length > 0) {
    const [atual, caminho] = fila.shift(); // shift() = popleft()

    for (const vizinho of (grafo[atual] || [])) {
      if (!visitados.has(vizinho)) {
        const novoCaminho = [...caminho, vizinho];

        if (vizinho === destino) {
          return { etapas: novoCaminho.length - 1, caminho: novoCaminho };
        }

        visitados.add(vizinho);
        fila.push([vizinho, novoCaminho]);
      }
    }
  }

  return { etapas: -1, caminho: [] }; // destino inalcançável
}

const origem  = "Twin Peaks";
const destino = "Golden Gate Bridge";

const { etapas, caminho } = bfsMenorCaminho(grafo, origem, destino);

console.log(`De: ${origem}`);
console.log(`Para: ${destino}`);
console.log(`Etapas (arestas): ${etapas}`);
console.log(`Caminho: ${caminho.join(" → ")}`);